### ISSUE TITLE
glm5_next: brief instruction at the head of a long prompt is never retrieved (GLM-5.3-Flash) — deterministic, not fixed by cache/chunking/DSA-dense

### ISSUE THREAD

**Issue by williamxie1989** [jundot/omlx#3248]
# glm5_next: brief instruction at the head of a long prompt is never retrieved (GLM-5.3-Flash) — deterministic, and not fixed by cache/chunking/DSA-dense

## Summary

On `GLM-5.3-Flash-oQ8-mtp` served through the mlx-vlm `glm5_next` path, a short
instruction placed at the **head** of a long prompt is reliably ignored, while the
**same prompt at the same total length** with the instruction moved to the **tail**
works every time. The model explicitly reports that it cannot see the user message.

This is reproducible byte-for-byte across server restarts, and — importantly — it is
**not** fixed by disabling the KV cache, chunked prefill, DSA sparse selection, or the
reasoning-effort template injection. I'm filing with the full exoneration table up front
so nobody spends a cycle on the obvious levers.

## Environment

- Mac Studio (512GB), omlx `0.6.3` @ `a20ec09d` (== `origin/main`, verified 0 commits behind)
- Model: `GLM-5.3-Flash-oQ8-mtp` (oQ8 affine, `group_size=64`, `bits=8`), 326.90GB
- `model_settings`: `enable_thinking=true`, `mtp_enabled=false`
- `settings`: `scheduler.chunked_prefill=true`, `cache.enabled=true`,
  `memory.prefill_min_chunk_tokens=32`
- Architecture: `num_hidden_layers=45`, **34 recurrent GDN/linear-attention layers**
  (`is_linear`) + **11 DSA full-attention layers** (layers 3,7,…,43),
  `index_topk=2048`, `index_kpool=4`, `index_kpool_always_select_tail=true`

## Minimal repro

```python
import json, urllib.request

BASE, KEY = "http://127.0.0.1:8015/v1/chat/completions", "1234"
MODEL = "GLM-5.3-Flash-oQ8-mtp"
SYS  = "You are a helpful coding assistant. Follow the user's instruction exactly.\n"
TASK = "TASK WORD: PURPLE-OTTER. Restate only the task word.\n"
FILL = "// neutral filler commentary line to pad context without special tokens\n"

def send(content, max_tokens=64):
    body = {"model": MODEL, "temperature": 0.0, "max_tokens": max_tokens,
            "messages": [{"role": "system", "content": SYS},
                        {"role": "user", "content": content}]}
    req = urllib.request.Request(BASE, json.dumps(body).encode(),
        {"Authorization": f"Bearer {KEY}", "Content-Type": "application/json"})
    d = json.loads(urllib.request.urlopen(req, timeout=900).read())
    m = (d.get("choices") or [{}])[0].get("message") or {}
    text = ((m.get("content") or "") + " " + (m.get("reasoning_content") or "")).strip()
    return d["usage"]["prompt_tokens"], ("PURPLE-OTTER" in text), text

pad = (FILL * 5000)[:40000]
print("HEAD", send(TASK + pad))   # *** LOST ***
print("TAIL", send(pad + TASK))   # SEEN
```

## Position-only flip (same total length)

| prompt_tokens | task @ HEAD | task @ TAIL |
|---|---|---|
| 6710 / 6711 | **LOST** | SEEN |
| 9209 / 9210 | **LOST** | SEEN |
| 11710 / 11711 | LOST | LOST |
| 15043 / 15043 | **LOST** | SEEN |

Only the instruction's *position* differs. Replies are byte-identical across restarts.

Model output at pt=6710 (HEAD):

> `We need answer to user. But no user message visible? Need inspect conversation: only syste…`

Model output at pt=11710 (HEAD) — note it **accurately describes the filler**, i.e.
everything *after* the ~11-token instruction is visible; only the leading instruction is lost:

> `The user has sent a message that is just "neutral filler commentary line" repeated many ti…`

So this is a **leading-span retrieval failure**, not wholesale prompt loss.

## What does NOT fix it (single-variable A/B, all re-run on the same build)

| Suspect | Test performed | Result |
|---|---|---|
| DSA sparse top-k dropping tokens | Added an env kill switch making `Glm5NextIndexer.__call__` return `None` → dense attention on all 11 FA layers. **Probe confirmed the switch fired** on every layer | still blind, byte-identical |
| Prefix cache | `cache.enabled=false` **and** removed the cache dir (verified 0B) | still blind, byte-identical |
| Stale SSD cache | moved the full 362GB `~/.omlx/cache` aside | still blind |
| Chunked prefill | `chunked_prefill=false` | byte-identical blind replies |
| Chat template / thinking | `reasoning_effort=low` / `high` / `clear_thinking=true` (outputs do change, so kwargs are consumed) | still blind |
| Chat template generally | raw `/v1/completions` | equally blind |
| Tokenizer / tokenisation | `apply_chat_template` + tokenize offline: token count **exactly 6710** == server-logged `prompt: 6710`; needle present at head of `input_ids` | correct |
| oQ conversion | parsed safetensors headers: all 11 indexers present with `wq_b`/`wk`/`weights_proj` + `index_kpool_compress_gate` `[128,4096]` / `_ape` `[4,128]`, Q8 affine g64 as mandated by `oq.py::_glm_indexer_q8_override` | correct |
| Native kernel build | all 6 fast symbols (`dsa_indexer_scores`, `dsa_topk_indices`, …) load on this box | available |
| Generic scheduler | `Qwen3.8-Flash-Next-oQ8e-mtp` serves 97K–101K agent traffic fine on the same box/config | fine |

Note the last row is not a counter-argument: real agent traffic puts the live
instruction at the **tail**, and the TAIL control passes at every length tested here.

## Why the architecture is implicated

With `index_topk=2048` / `index_kpool=4`, the production sparse path attends to
~2048 of e.g. 12625 prompt tokens (~16%). Combined with 34/45 layers being recurrent
GDN state (fixed-size, strong recency bias), a brief instruction buried at the head of
a long, highly repetitive context is the worst case for this layout.

The confusing part — and the reason I'm filing rather than closing this myself — is
that **forcing dense attention on the 11 full-attention layers does not recover the
head**. If the FA layers hold correct KV for all positions and still cannot surface a
10-token leading instruction, something upstream of attention selection is off. I
verified cache accounting is exact (`6144 + 565 + 1 == 6710`), so tokens do reach the
FA layers.

Also worth a look: the model ships `chat_template: None` in `tokenizer_config.json` and
relies on `chat_template.jinja`, whose lines 2–3 emit an unconditional
"Reasoning Effort: Max" system block **before** the real system block (defaulting to
`max` whenever `reasoning_effort` is not `low`/`high`). The rendered head is therefore:
`[gMASK]` + `<sop>` + im_start/system role token + `Reasoning Effort: Max` +
im_start/system role token + real system prompt + im_start/user + … . The duplicated
system role token at the very head looks unusual. It did not cause this symptom in my
tests (`reasoning_effort=low`/`high` change the output but not the blindness), but it
may be worth confirming against the reference implementation.

## Workaround currently in use

Ordering the messages so the live instruction is the **last** user turn (moving
runtime-context / skill catalogs into a system message ahead of it) makes every tested
length succeed. Repetitive catalog content also appears to matter a lot.

## Questions

1. Is head-span retrieval at >6K known-degraded for `glm5_next`, or is there a config
   (`index_topk`, `index_kpool`, `clear_thinking`, `bypass_short`) intended for it?
2. Should `bypass_short` (dense fallback below `index_topk`) be measured in tokens or
   pooled windows? It reads as pool units here, which would make the dense window 8192
   tokens rather than 2048.
3. Is the duplicated `system` role header in `chat_template.jinja` lines 2–3 correct?

Related but distinct: #2490 (tail loss, fixed via DeepSeek DSpark row-wise attention —
does not cover glm5_next), #2394 (~6k truncation; not this, prompt length is honoured in
full here), #3226 / PR #3227 (Metal buffer-count crash; orthogonal to this symptom).

**Comment by williamxie1989** [comment]
## Follow-up: length threshold + full component exoneration (2026-08-28)

Ran a finer sweep and probed every omlx-side component. Summary:

**Length threshold (task at HEAD, filler after):**
| prompt_tokens | ~43 | ~376 | ~710 | ~1046 | ~1377 | ~2710 | ~6710 |
|---|---|---|---|---|---|---|---|
| HEAD | SEEN | SEEN | SEEN | LOST | LOST | LOST | LOST |

The transition is between ~710 and ~1046 tokens. Holds for **both** repetitive and
diverse (non-repeating) filler, so it is not repetition-specific — it is a general
length-dependent head/middle-retrieval failure.

**Every omlx-side component probed and found correct:**
- **FA mask**: `create_attention_mask` row-sums are exact — first query sees
  `offset+1` keys, last query sees `offset+S`. No head masking.
- **kv_latent integrity**: FA-layer latent norms at the head are healthy (7–18),
  comparable to mid/tail. The head representation reaches the FA layers intact.
- **Dense FA**: forcing `Glm5NextIndexer.__call__ → None` (dense attention on all 11
  FA layers, probe-confirmed) does **not** recover the head.
- **GDN kernel**: `gated_delta.py` is a standard, correct Gated DeltaNet recurrence
  (forget-decay → read → delta-write → output). No chunk drop, no state bug.
- **GDN zeroed**: zeroing the 34 GDN layers' output (cache updates preserved) makes
  the model emit incoherent tokens — confirming the GDN layers carry most of the
  representational capacity, and that the FA layers alone cannot surface the head.
- Tokenizer (count exact, needle present), KV/SSD cache (off → still blind), chunked
  prefill (off → identical), oQ indexer quantization (shapes/spec exact), native
  kernels (all symbols load) — all previously exonerated.

**Working hypothesis:** with 34/45 layers being fixed-state GDN (recency-biased
linear attention) and only 11 full-attention layers, the recurrent state saturates
with recent content past ~1k tokens and the FA layers' long-range signal is
overwhelmed in the residual stream. Dense FA failing to recover the head indicates
the limitation is in the model's layer mixing / GDN dominance, not in any omlx
mask, cache, quantization, or kernel path.

Open question for maintainers: is this expected for `glm5_next` at this depth, or is
there a known config (GDN forget-gate, `index_topk`, `hc_mult` mixing) that restores
head/middle retrieval? Happy to provide the probe scripts.

**Comment by williamxie1989** [comment]
## Root cause found + fix (2026-08-29)

This is **not** a model-architecture limitation — it is a concrete omlx code bug,
now fixed and verified.

### Root cause
`omlx/patches/mlx_vlm_glm5_next_compat/vendor/mlx_vlm/models/glm5_next/linear.py`
routes the glm5_next projections (`q_a_proj`, `q_b_proj`, `kv_a_proj_with_mqa`,
`o_proj`, MLP/MoE) through `linear_forward()`, which — once the token count
reaches `min_tokens = 1024` for 8-bit weights — dispatches to the **Qwen3.5
native affine-qmm tile** `qwen35_q8_affine_qmm_t` (from
`omlx.custom_kernels.qwen35_prefill`).

That tile returns **incorrect results** for glm5_next's 8-bit affine weights
(the `_is_supported_affine_linear` gate does not catch it). The corrupted
q/kv projections break long-context attention: a brief instruction at the head
of the prompt is not followed once the prompt exceeds ~1k tokens, while the
identical prompt works below the threshold.

**The 1024 threshold is the smoking gun** — it exactly matches the observed
SEEN→LOST boundary (SEEN at ~710 prompt tokens, LOST from ~1046 on).

### Why it looked like a "second bug" at longer lengths
The buggy tile also wrote **corrupted KV values into the tiered SSD cold
cache**, which persisted across restarts and kept failing even after the tile
was bypassed. From a clean cache the single fix resolves every length.

### Fix
`glm5_next/linear.py` is glm5_next-specific, so `linear_forward()` now routes
through standard `mx.quantized_matmul` unconditionally (correct at every
length). This matches the known-good `feat/glm5-next` implementation, which
never used the native tile.

### Verification (GLM-5.3-Flash-oQ8-mtp, task word at HEAD)
| prompt_tokens | 43 | 710 | 1377 | 2710 | 6710 |
|---|---|---|---|---|---|
| before (main) | SEEN | SEEN | LOST | LOST | LOST |
| after (fix, clean cache) | SEEN | SEEN | SEEN | SEEN | SEEN |

Repro: put a one-line instruction at the head of a >1k-token prompt and ask the
model to restate it; it fails on main and succeeds with the fix. Branch with the
fix: `fix/glm5-next-native-qmm`.

**Comment by SomeOddCodeGuy** [comment]
Popped on here during my lunch break and saw this, so just adding a quick comment.

I'm not sure if this is the same issue I saw last night, but this post lines up perfectly with a frustration I was having with GLM 5.3 flash and could not figure out what in the world was going on.

I had a ~1066 token prompt that I was running through omlx 0.6.3 with Pipenetwork's `GLM-5.3-Flash-MLX-8bit`. I was hitting the API directly with curl for testing because I thought maybe my middleware or front-end was the cause.

* Scenario 1: I send just a single, simple, user prompt of "What is Final Fantasy XIV?" **It answers fine**
* Scenario 2: I send my normal lengthy system prompt + normal length user prompt with the same question? **Hallucinated responses talking about everything from C++ to counting words in a "test_doc.txt" that never existed.**
* Scenario 3: I send same lengthy system prompt, and remove a couple paragraphs from user prompt? **It answers fine**

Now here is where it gets weird:
* Redid Scenario 3 multiple times: 3 out of 3 attempts **successfully** answered "What is Final Fantasy XIV?" without issue
* I only added the word `genre` and nothing else: "What genre is Final Fantasy XIV?" 3 out of 3 **failures** due only to adding the word genre. Every response was saying my messages were blank or I forgot to send a message.

Reproduced all of the above by both sending unsloth guide's recommended samplers with reasoning effort max, and also separately trying nothing else except for `temperature: 0`. Weirdness no matter what samplers I went with. I also disabled prefix cache completely, restarted omlx multiple times, and did streaming and non-streaming requests.

The whole thing was so weird what I didn't even know where to report it to, much less what to report. I've never seen this before. I thought maybe the tokenizer was messed up, but then I used "genre" in another short user prompt and it worked fine, so I was just stumped. I figured either omlx had an issue or the model did. But when I saw your post, I thought "Its not identical, but sure is close", so I figured I'd mention that here.

**Comment by williamxie1989** [comment]
> Popped on here during my lunch break and saw this, so just adding a quick comment.
> 
> I'm not sure if this is the same issue I saw last night, but this post lines up perfectly with a frustration I was having with GLM 5.3 flash and could not figure out what in the world was going on.
> 
> I had a ~1066 token prompt that I was running through omlx 0.6.3 with Pipenetwork's `GLM-5.3-Flash-MLX-8bit`. I was hitting the API directly with curl for testing because I thought maybe my middleware or front-end was the cause.
> 
> * Scenario 1: I send just a single, simple, user prompt of "What is Final Fantasy XIV?" **It answers fine**
> * Scenario 2: I send my normal lengthy system prompt + normal length user prompt with the same question? **Hallucinated responses talking about everything from C++ to counting words in a "test_doc.txt" that never existed.**
> * Scenario 3: I send same lengthy system prompt, and remove a couple paragraphs from user prompt? **It answers fine**
> 
> Now here is where it gets weird:
> 
> * Redid Scenario 3 multiple times: 3 out of 3 attempts **successfully** answered "What is Final Fantasy XIV?" without issue
> * I only added the word `genre` and nothing else: "What genre is Final Fantasy XIV?" 3 out of 3 **failures** due only to adding the word genre. Every response was saying my messages were blank or I forgot to send a message.
> 
> Reproduced all of the above by both sending unsloth guide's recommended samplers with reasoning effort max, and also separately trying nothing else except for `temperature: 0`. Weirdness no matter what samplers I went with. I also disabled prefix cache completely, restarted omlx multiple times, and did streaming and non-streaming requests.
> 
> The whole thing was so weird what I didn't even know where to report it to, much less what to report. I've never seen this before. I thought maybe the tokenizer was messed up, but then I used "genre" in another short user prompt and it worked fine, so I was just stumped. I figured either omlx had an issue or the model did. But when I saw your post, I thought "Its not identical, but sure is close", so I figured I'd mention that here.

@SomeOddCodeGuy I have identified the issue and fixed it. I've already submitted a PR https://github.com/jundot/omlx/pull/3266. Perhaps you could cherry-pick it and see if it works for you.Additionally, don’t forget to clear the SSD cache after applying the changes

**Comment by SomeOddCodeGuy** [comment]
Sounds great. When I get off work I'll give it a shot!

**Comment by mightyC1** [comment]
Independently reproduced and root-caused (flip at exactly 1024 prompt tokens = the q8 `min_tokens` routing gate). The tile is actually bit-exact on contiguous inputs; the corruption comes specifically from the non-contiguous `mx.split` views (`fa_o`/`ga_o` → KDA forget/output gates) passing a lazy-flags contiguity check. Details, a 10-line weight-free synthetic repro, and a minimal perf-preserving alternative fix posted in #3266.

**Comment by SomeOddCodeGuy** [comment]
> [@SomeOddCodeGuy](https://github.com/SomeOddCodeGuy) I have identified the issue and fixed it. I've already submitted a PR [#3266](https://github.com/jundot/omlx/pull/3266). Perhaps you could cherry-pick it and see if it works for you.Additionally, don’t forget to clear the SSD cache after applying the changes

I can confirm that a build using your branch has resolved the issue for me. All my previous failing tests are now working as expected.

**Comment by jundot** [comment]
@williamxie1989 @SomeOddCodeGuy @mightyC1 Thanks for the detailed reproductions and the work narrowing this beyond cache, chunking, and attention selection. I merged #3275; the root cause was the `fa_o`/`ga_o` split views reaching a native QMM tile that reads its input as row-major contiguous. On the merged `main`, all 22 GLM compatibility tests and the real-model long-prompt smoke test pass. If the same prefix was cached by an affected build, please clear that stale GLM cache entry once after updating so it is recomputed through the fixed path. I’ll close this as fixed.
